### ISSUE TITLE
neovim integration?

### ISSUE THREAD

**Issue by prabirshrestha** [VSCodeVim/Vim#18]
Are there any plans for neovim integration?

**Comment by jpoon** [comment]
As I understand, the neovim project is refactoring vim core itself. What would be involved with integrating VSCode with neovim?

**Comment by snowiow** [comment]
As far as I know, neovim offers the RPC protocol for communication. So a gui can communicate with the neovim core for example. Here are some [example projects](https://github.com/neovim/neovim/wiki/Related-projects). Here is also a [small overview](https://github.com/neovim/neovim/wiki/Plugin-UI-architecture#ui-programs) on how this communication could look like. But I don't have a clue about anything beneath this theory.

**Comment by jpoon** [comment]
Thanks for sharing! Whoa, that is actually really amazing. I'm curious as to its current state; of the example projects, none of them looked very complete. Nonetheless, looks super promising.

**Comment by jordwalke** [comment]
I would seriously consider taking a look at this existing electron-integrated neovim implementation: https://github.com/coolwanglu/neovim-e

It works really well. The challenge with integrating into VSCode is that neovim currently only exposes raw character/line draw command (this will be improved over time). VSCode's API likely will want something more rich than this. This is why that neovim-electron project uses canvas to paint everything.
However, I would still recommend *starting* with an interpretation of NeoVim's raw/unstructured commands - counting on NeoVim to deliver more structured information in the future (or counting on the fact that you can extract whatever meaning you need from the unstructured commands with enough reliability).
Comparing the time it would take to interpret the unstructured message (or wait for NeoVim to deliver more structure) against the time/effort it takes to replicate all of Vim's nuances correctly - I would say that integrating with NeoVim might get you to the end goal more quickly.

(For context, I'm a Vim emulator connoisseur, and file countless issues on various editor plugins that try to get the Vim feel right - in my experience, it takes a very very long time - and not a single one has achieved Vim's core editing experience. With *much* due respect to Vintageous (which does a *very* good job in general), even *it* doesn't replicate the undo/redo cursor position correctly (which is a critical part of the Vim editing experience) (but in that case because it's fundamentally limited by Sublime's API).

**Comment by guillermooo** [comment]
How stable is Neovim on all three platforms?

I'm against adopting Neovim wholesale at this point. My suggestion is that we continue with a port of Vim in TS without taking it too far. At the same time, we can explore certain functionality using Neovim (vimrc support, for example, if it provides that), and then converge on the choice that makes more sense: pure TS port or Neovim back-end.

Also, we'll need some TS infrastructure anyway, primarily for keyboard handling, so we can implement a handful of commands in TS, see where we get and then adjust our approach wrt Neovim.

**Comment by jordwalke** [comment]
> How stable is Neovim on all three platforms?

I'm not sure, I've only tried it on Mac OS, but it's forked from Vim.

>  Also, we'll need some TS infrastructure anyway, primarily for keyboard handling, so we can implement a handful of commands in TS, see where we get and then adjust our approach wrt Neovim.

Yeah, don't someone like me stop you - seems like good stuff will come out of this! I'm just giving my perspective as a *consumer* of Vim emulators. A Vim emulator that only gets 98% right is almost worthless - no one ends up using them. It's the *small* subtleties that are the worst. Please consider it a great compliment to your fantastic work on Vintageous that I consider it "faithful enough to use". I just thought that there might be a faster path to getting there for `VSCodeVim` (I'm recalling the massive amount of work that I've seen you do over the last several years - my how time flies :) )

**Comment by tmandry** [comment]
FWIW, all reports I've heard say neovim is very stable, thanks in part to its dramatically expanded body of tests. A few reports are [here](https://news.ycombinator.com/item?id=10488188). I do agree completely with @jordwalke's point of view here. I personally think it's only a matter of time before integrations like this replace emulations, in editors where it's practical to integrate.

**Comment by tmandry** [comment]
Starting points for anyone who is interested in looking into this:

- [Plugin UI architecture](https://github.com/neovim/neovim/wiki/Plugin-UI-architecture)
- [API](https://github.com/neovim/node-client/blob/master/index.d.ts)

Also you might want to read tarruda's post about his planned [Smart UI protocol](http://tarruda.github.io/articles/neovim-smart-ui-protocol/) to get a sense of what's coming in the future.

**Comment by jordwalke** [comment]
That "Smart UI Protocol" is exactly what I had in mind.

In the mean time, I would even be happy with a *canvas* rendering of Vim inside of VSCode (which is almost trivial to accomplish thanks to neovim-e (electron example)).

**Comment by drewcassidy** [comment]
> In the mean time, I would even be happy with a canvas rendering of Vim inside of VSCode (which is almost trivial to accomplish thanks to neovim-e (electron example)).

:100: This would be awesome, especially if i can still use all of VSCode's debugging features. This might even be the right thing to do given the current limited extension API.

**Comment by aakashsigdel** [comment]
From my experience I can tell that neovim is very stable in mac OSX and in linux. And I would love to see it being integrated with vscode.

**Comment by songgao** [comment]
Another happy `neovim` user here. I have `alias vim=nvim` and never had any serious (if any at all) problems. The team is very active and responsive to issue reports. Not sure about API stability though.

**Comment by owais** [comment]
FWIW, I've been using neovim-dot-app (https://github.com/rogual/neovim-dot-app) on OS X on production for a couple of months at least and it works really really well. Only minor issue I've seen so far is that the re-draws are a bit slow when scrolling too fast.

**Comment by johnfn** [comment]
This is still viable as an extension, but probably isn't going to be ours, given the amount of stuff we've already done... 😛

**Comment by johnfn** [comment]
To the people that are thumbsing this up (and thumbsing down my previous comment): What particular, specific features of Neovim are you missing from VSCodeVim? (e.g. "plugins" is not very helpful, but "surround.vim" is.)

**Comment by justinmk** [comment]
@johnfn I didn't vote, but some Vim features are missing from all vim emulations: ex commands (there are hundreds of them), vim regex, jumplist, quickfix list, macros, marks (the non-obvious ones), vim's undo behavior, spellcheck (it's powerful and unique), vim completion (it's powerful and unique), viminfo (called "shared data" in Neovim). And lots of subtle details.

**Comment by johnfn** [comment]
@justinmk I know that there are lots of things missing from VSCodeVim, but I'm looking for the specific features that are missing that inspired you guys to upvote this thread, so we can target our efforts in that way. 

(There's no real point to work on, say, the quickfix list, if no one derives benefit from having it...)

**Comment by jordwalke** [comment]
My support of true NeoVim integration has nothing to do with a particular feature but more so a general philosophy I've gained after trying and watching dozens of vim key bindings implementations miss very basic features. There are many nuanced pieces to vim's interaction language that it is inevitable that you will miss many of them and not even realize it. The amount of time required to integrate vim directly is a fraction of the time required to reimplement even the most basic vim interactions with high accuracy. I have only seen one plugin ever get the basics right and I've tried about twenty different attempted reimplementations thoroughly. It typically only takes two minutes of usage to spot a major defiiciency/bug even for projects that have been invested in for years.

**Comment by jordwalke** [comment]
Unless you are going to invest massively in a VSCode vim emulator from scratch, or start with vim-mode-Plus and adapt it to VSCode, I do not have high hopes that the resulting experience will win over any true vim fans. I want to believe. I really do. I've just been burned too many times. I highly recommend embedding NeoVim or vim directly.

**Comment by johnfn** [comment]
@jordwalke Have you used VSCodeVIm? Could you please point out the areas in which it is deficient?

I mean, if you guys were like "I really desperately need 100% compatible viml evaluation" then that is complex enough that I would consider a NeoVim integration. But if you were just like "it's frustrating that the cursor ends up in the wrong place when I do [random command]," I mean, that's something we can work with.

**Comment by justinmk** [comment]
@johnfn For me personally, it's easier to say what I want from an IDE or higher-level editor, rather than enumerate the things I want from vim. 

To be clear, I want:

- vim (neovim) running in a single tab in atom
    - File navigation is done by vim. I don't need atom tabs, splits etc.
- The ability to map vim keybindings to atom API functions. Examples:
    - Bind a key to invoke some refactoring capability of atom (assuming that's exposed as an API)
    - Bind a key that invokes atom's "go to class" navigator. When I make a choice, it opens the file in the embedded neovim instance. (Neovim API makes this trivial)
- (optional) some way to invoke the IDE (atom) context menu at the cursor position, or do other actions associated with the cursor position

I don't care about atom's search/replace, find-in-files, code completion, text effects, or anything involving text-editing.

**Comment by johnfn** [comment]
> File navigation is done by vim. I don't need VSCode tabs, splits etc.

When you say file navigation, what exactly do you mean? 

> Bind a key to invoke some refactoring capability of VSCode (assuming that's exposed as an API)
> Bind a key that invokes VSCode's "go to type" navigator. When I make a choice, it opens the file in the embedded neovim instance. (Neovim API makes this trivial)

We already have these two things. :)

**Comment by justinmk** [comment]
> When you say file navigation, what exactly do you mean?

Finding and loading files. I mentioned that just to avoid discussion about hooking into various VSCode dialogs, menus, etc. 

> We already have these two things. :)

Well, the embedded instance of nvim is missing. That's the part that eliminates unending quest for emulation, and instead unlocks the most compelling features of VSCode (refactoring and navigation) for people who like vim. Vim has very poor support for refactoring and semantic navigation ("find references", "go to type"), those are the parts vim users want.

MS Language Server Protocol helps a lot with this, but even then, VSCode has so much of that already built-in.

**Comment by johnfn** [comment]
> Finding and loading files. I mentioned that just to avoid discussion about hooking into various VSCode dialogs, menus, etc.

What exactly are the commands you are talking about here? I assume you aren't just talking about `:e`. 

> That's the part that eliminates unending quest for emulation

The quest isn't unending if you guys can tell me what specifically you want 😉  It would be very useful for me to understand what features an embedded nvim has that would make it superior to our implementation.

**Comment by justinmk** [comment]
It's not necessarily superior. It's just that one has [everything](https://neovim.io/doc/user/) (for better or worse) and the other is implementing a subset based on what's most popular. Different users want different parts. 

I listed [several features](https://github.com/VSCodeVim/Vim/issues/18#issuecomment-272691725). For me, Vim-style completion (`ctrl-n`, `ctrl-x-...`, ...) is a big one. Re-implementing Vim-style completion--just that single feature--will be more work than embedding Neovim.  And probably few people care about Vim-style completion so it's not going to get re-implemented anyway. 

Another example: the `'[` and `']` marks. VsVim (for Visual Studio) doesn't have those, I requested it years ago. Curious to know if VSCodeVim has those :)

**Comment by johnfn** [comment]
> It's not necessarily superior. It's just that one has everything (for better or worse) and the other is implementing a subset based on what's most popular. Different users want different parts.

Right, and I'm curious about which parts are the ones that you guys want. 

From your list:

* We have regex (but it is the JS regex)
* Jumplists are actually part of VSCode itself (cmd + and cmd -)
* Quickfix is actually part of VSCode too
* VSCodeVim has macros
* VSCodeVim has (most) marks
* VSCode provides completion (it's not a direct imitation of Vim's completion, obviously)
* Spell check is provided by a [separate plugin](https://github.com/Microsoft/vscode-spell-check)
* .viminfo is something we definitely don't have. 

So hopefully now you understand why I continue to ask for specific things that are missing. To me, it feels like large swaths of the functionality that you have asked for exists. But of course I would totally understand if you said something like "the regex engine isn't faithful to Vim" or something. 

We definitely don't have Vim-style completion. What about vim style completion do you find superior to VSCode completion? I always felt like VSCode was vastly superior to Vim in that area, so I'd be curious to know.

**Comment by justinmk** [comment]
> We have regex (but it is the JS regex)

Again, it's not vim regex, which has atoms like `\%#` and `\zs`.

> Jumplists are actually part of VSCode itself (cmd + and cmd -)

Not the same at all. Nor the other things.

> So hopefully now you understand why I continue to ask for specific things that are missing.

I didn't mention "vim spellcheck" thinking that VSCode didn't have spellcheck. I'm aware that every single program on earth has spellcheck. I mentioned "vim spellcheck" specifically because it has unique features. 

And indeed, I just confirmed that the `'[` and `']` marks don't work in VSCodeVim.

**Comment by jordwalke** [comment]
> But if you were just like "it's frustrating that the cursor ends up in the wrong place when I do ..

> Have you used VSCodeVIm? 

Yes, I have tried every Vim plugin in existence that I've been able to find including VSCodeVim. Almost all of them get critical things wrong including undo/redo/cursor positioning, repeat etc. I used to report all the issues I found but developers of Vim emulators almost always are not interested in recreating Vim after revealing how difficult all of these edge cases are to cover.

I can try VSCodeVim once again, and report all the errors that I find, but I do suspect it's not as productive as just integrating NeoVim. (I will still do it though). The only reason this is my current stance is due to my very extensive experience in testing existing attempts at doing wht VSCodeVim is doing.

Unlike others here, I really like what VSCode has to offer in terms of file search, command pallet, git integration, plugin loading, extensibility, the language service API. But text editing is a solved problem and one that I believe, quite frankly, is done better in Vim than any other editor (if you don't like modal editing, disable Vim's modal editing and just use it like a normal editor). VSCode offers so much outside of the raw act of text editing that the best editor would be a combination of the VSCode ecosystem/UI with the language service protocol, but delegating the raw text editing to Vim. Vim even has unto *tree* capabilities which would be very very difficult to recreate in an editor.

Hope I don't sound too skeptical, I really do want to help VSCode achieve the best Vim editing mode possible and after trying several other attempts, I've come to the conviction that embedding vim or neovim is the fastest way to get there. In fact, I don't even know why NeoVim couldn't power VSCode's primary editor - using the MS Language Server protocol and a Vim Shim for it. (One reason to *not* use Vim for the primary editing engine of VSCode is that you could not easily run it on the web (though people have built Vim for web using Emscripten)).

**Comment by jordwalke** [comment]
Oh, and could you imagine how fast VSCode's load time would be if the entire text editing framework could be replaced with the *native* Vim engine that removes virtually *all* startup time (we pay for all that JS parsing/loading/JIT at startup time). How much of VSCode's startup time is due to loading JS files and how much of that is attributed to the JS implemented text editor?

**Comment by johnfn** [comment]
@justinmk Can you please please go into details on why the features that you like from Vim are different than the features that exist in VSCodeVim/the VSCode ecosystem? I genuinely don't know and I would like to understand.

**Comment by justinmk** [comment]
That's a very time-consuming request.

**Comment by johnfn** [comment]
@justinmk In order to make a good implementation of Vim in VSCode, I need to understand what the usage patterns of Vim users look like. 

When you say "there are many features that are deficient" but don't actually detail what the deficiencies are, I have no information upon which I can act to choose the best way to improve the plugin. I could go learn about them, but how would I know that I was capturing the correct use cases?

**Comment by jordwalke** [comment]
I'll mention one that I haven't confirmed in a while - but I'm assuming VSCodeVim doesn't implement it correctly: When redoing a change, the cursor should be positioned at the beginning of the redone change. Nearly every vim emulating plugin gets this wrong (vmp is the only one that doesn't). I've tried all of them.
Also, VSCodeVim doesn't implement an undo tree, like Vim does - it uses an undo stack (VSCode's) IIRC.

**Comment by johnfn** [comment]
@jordwalke I'm pretty sure we get the redo position right. Go ahead and try :P

**Comment by johnfn** [comment]
@jordwalke I know that it seems like NeoVim is probably not as hard as implementing the remainder of those features. My current perspective a NeoVim port is the following:

1. Actually correctly implementing a NeoVim integration would be a magnitude of work probably equal to or exceeding the amount of work we have already done. Consider:

* How multiple cursor support should be done with a NeoVim integration
* Synchronizing registers with VSCode's clipboard
* Synchronizing buffers with VSCode's files
* Integrating VSCode snippets with NeoVim
* Making NeoVim play well with other VSCode plugins (How would NeoVim understand if a different plugin inserted text somewhere, bypassing our keybindings? How would NeoVim understand when to allow VSCode to process keystrokes, vs when it should process keystrokes?)
* Mapping the contents of NeoVim's various input boxes to somewhere in VSCodeVim
* Mapping NeoVim's folds to VSCode's folds.

To be clear, I'm not saying it's a bad idea, or that I won't do it. Just that it may be more work than it appears.

2. I believe that VSCodeVim currently captures the 90%+ use case for most users, and that most of the crucial things it lacks are relatively simple to add. The reason I believe this is because when I look over the [most popular issues on our issues page](https://github.com/VSCodeVim/Vim/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc), they generally fall into one of a few categories:

* Things that VSCode currently doesn't allow us to do
* Minor bugs
* Really simple features

Like, if I went to that page and I saw that "Undo Tree History" was our most popular issue with 500 thumbs up, I would probably start considering a NeoVim port. 😉  But that hasn't proven to be the case.

As usual, I am open to changing my mind. I'd need to be convinced that a NeoVim port would be easier (1) than just implementing the missing features to round out the 90% case for the users of VSCodeVim (2).

**Comment by jordwalke** [comment]
I love your optimism!

I just tried the cursor position on undo/redo again, and it looks better than most other attempts, but still not "correct". See this example:

Enter:
```
aaaa
bbbb
cccc
dddd
```

Visually select from the first `b` to the last `c`. Hit `c` to change the text, and replace the two lines with:

```
qqqq
qqqq
```

So that the final text is:

```
aaaa
qqqq
qqq
dddd
```

The hit escape, and undo/redo that change repeatedly. The cursor should remain on the first `q` location as you undo and redo that change. It remains at the end.

I only played with this for a couple minutes, so I could probably find a few more issues if you like.

**Comment by jordwalke** [comment]
The other thing about all the challenges of integrating vim/neovim you mentioned - is that even if those things take time and don't start off perfect, it's *okay*. Vim users can switch if only a fraction of the VSCode vim integrations are supported because hey, it's exactly what they're used to *and* they get a handful of VSCode integrations. Some of the ones you mentioned aren't that hard, others are certainly more difficult, and others might not be that important afterall, when you can create a bridge between the VSCode config/feature and an existing Vim plugin. For example, imagine creating a conversion tool from VSCode snippets to Ultisnips (which is the standard in Vim snippets plugins).

**Comment by jordwalke** [comment]
To be clear, I do believe you could simply fix all of these issues (and make it more polished (the cursor selection appears visually to be a little flaky (but that's just styling))), but I only wanted to point out that I've seen this become an all-consuming task for developers. Some (like vmp) have been able to pull it off.

**Comment by johnfn** [comment]
I totally understand where you guys are coming from, and you've convinced me that, given unlimited time, I would definitely try my hand at a NeoVim port. Unfortunately, I barely have any time at all right now, so it seems like a more effective use of my resources to spend a few hours getting VSCodeVim from 94% to 97% than the same number of hours getting VSCodeNeoVim from 0% to ???%. 

I mean hey, if you guys wanna throw money at me to pursue NeoVim, I won't say no :P As it is though, I'm just trying to pursue the thing that benefits the most people with the limited amount of time that I can spend on this plugin.

**Comment by justinmk** [comment]
No problem. My only goal was to answer [your question](https://github.com/VSCodeVim/Vim/issues/18#issuecomment-272688827).

**Comment by jordwalke** [comment]
*Totally* understandable and thanks for your work so far. You might also consider taking a look at `vmp` from Atom. Perhaps you could join forces if you want to go this direction and somehow help make `vmp` generalizable to more editors.

**Comment by justinmk** [comment]
> then send NeoVim a command string, NeoVim will send back a 80x200 2D array telling you what's on each position.

Some widgets are being externalized. Currently in Nvim 0.1.7 the popupmenu is externalized, so UIs can render it however they like. We're close to externalizing other widgets like wildmenu, statusline, tabline, and cursorline.

> That's impractical for integrating to VSCode, for obvious perf reason, 

What perf reason? UIs should (and do) batch the redraws.


> Text diff. Say I press di}, just tell me "line 20-40 is affected, apply this diff to line 20-40 in VSCode, then you get what you get in NeoVim".

https://github.com/neovim/neovim/pull/5269

> I remember seeing some progress in NeoVim 0.2 release blog draft 

https://neovim.io/news/2016/11/

> Also remember some work was being done in NyaoVim

Also check out: https://github.com/extr0py/oni

**Comment by octref** [comment]
Thanks. I'm subscribing to #5269, and I'll try again once it's in NeoVim.

Oni looks cool, but I'm worried how far it can go.

**Comment by v3ss0n** [comment]
Oni is still quite unstable , can't use for daily work.
Oni should just do a plugin for VSCode , isntead of developing whole ide themselves.

**Comment by johnfn** [comment]
I've been monitoring other attempts at integrating NeoVim, but they appear to be blocked on very simple issues. e.g. `dw` does not currently work, apparently: https://github.com/neovim/neovim/issues/6166

I will continue to monitor these plugins to see when NeoVim gets into a usable state!

**Comment by jordwalke** [comment]
That's really strange - I've had no problem with `dw` in NeoVim. Is it just a particular use of that protocol they're using that causes this?

**Comment by johnfn** [comment]
@jordwalke Yeah, not to say that dw doesn't work in NeoVim - that would be insane. I believe it has something to do with the remote msgpack API. 

Like I said before, if someone wants to fund me, I'll totally get started on a NeoVim plugin for VSCode today. 😁

**Comment by jordwalke** [comment]
I'm absolutely certain you would raise a great chunk of change on KickStarter.
Have you heard of other software projects using a similar model?

NeoVim might be a nice project for that model because there's clearly a few simple steps along the way.

- Turn each editor tab into a simple canvas that just uses NeoVim's "dumb" protocol, and integrate it with the standard VSCode hooks for opening files etc.
- Begin adding "smart" UI protocol features from neovim.
- Extra polish.

**Comment by justinmk** [comment]
The "dw" RPC bug is a high priority. 

[actualvim](https://github.com/lunixbochs/actualvim) recently switched to Neovim, and the result is impressive (if you have Sublime Text 3, try it). It's very similar to the model I tried to describe above.

It's about 1000 lines of code, and it blows away any other "vim mode".

**Comment by johnfn** [comment]
Ah @justinmk I didn't realize you were one of the main contributors to NeoVim! That explains some things...

**Comment by octref** [comment]
> Turn each editor tab into a simple canvas that just uses NeoVim's "dumb" protocol

That alone is impossible as VSCode doesn't allow extensions to touch its DOM.

**Comment by 3cooper** [comment]
I've been keeping track of this thread for sometime.  I'd support a kickstarter to use  neovim as the backend for the plugin.

**Comment by tmandry** [comment]
Same, I've been supporting neovim since the beginning for exactly this
application.

**Comment by clee** [comment]
@johnfn One missing feature that I have started to heavily use in "real" vim (and neovim) is register support, which does not appear to work at all in VSCodeVim.

For anybody who doesn't use it, you can type `"` to enter register-selection mode, then any of the `[0-9a-z]` characters (and possibly more, but I've only tested with those) to select a register to store/pull buffers. So, an example would be, `"kdd` to delete the current line and store it in the `k` register, and then later you can use `"kp` to paste whatever you stored there. When you're working with multiple chunks of code, it's really useful.

**Comment by johnfn** [comment]
@clee Register support 100% works in VSCodeVim, aside from a few edge cases. Please open up an issue describing your problems!

**Comment by clee** [comment]
@johnfn When I tried it, it seemed like any non-register `yy` or `dd` command killed my register buffers. But I can definitely give it a shot again!

**Comment by Chillee** [comment]
I've gotten a proof of concept working here! https://github.com/VSCodeVim/Vim/pull/1725

It should support all text-editing Ex commands.

**Comment by Chillee** [comment]
As we've started adding neovim integration, please leave all feedback in this issue: https://github.com/VSCodeVim/Vim/issues/1735

**Comment by yetone** [comment]
https://github.com/neovim/neovim/wiki/Related-projects#gui

**Comment by hyiltiz** [comment]
Maybe integration with an even smaller codebase like `vis` would be ideal; `nvim` supports `vim`'s plugins and vimscripts, which doesn't quite make sense for VSCode.

**Comment by justinmk** [comment]
> `nvim` supports `vim`'s plugins and vimscripts, which doesn't quite make sense for VSCode.

Depends on the plugins. Many of them make sense.

In any case, this is a solved problem: https://github.com/asvetliakov/vscode-neovim
